bin/cheribsdtest/cheribsdtest_colocation.c

```diff
@@ -0,0 +1,179 @@
+ *
+ * This software was developed by SRI International and the University of
+ * Cambridge Computer Laboratory (Department of Computer Science and
+ * Technology) under Defense Advanced Research Projects Agency (DARPA)
+ * Contract No. HR001122C0110 ("ETC").
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/cdefs.h>
+
+#include <sys/param.h>
+#include <sys/socket.h>
+#include <sys/sysctl.h>
+#include <sys/queue.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+
+#include <err.h>
+#include <errno.h>
+#include <kvm.h>
+#include <libprocstat.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <sysexits.h>
+#include <unistd.h>
+
+#include "cheribsdtest.h"
+
+static bool
+is_colocated_with_parent(void)
+{
+	struct procstat *psp;
+	struct kinfo_proc *kipp;
+	struct kinfo_vmentry *kivp;
+	pid_t pid, ppid;
+	uint pcnt, vmcnt;
+	bool found_self, found_parent;
+
+	pid = getpid();
+	ppid = getppid();
+
+	psp = procstat_open_sysctl();
+	if (psp == NULL)
+		err(EX_OSERR, "procstat_open_sysctl");
+
+	kipp = procstat_getprocs(psp, KERN_PROC_PID, getpid(), &pcnt);
+	if (kipp == NULL)
+		err(EX_OSERR, "procstat_getprocs");
+	if (pcnt != 1)
+		warnx("got %d processes", pcnt);
+
+	kivp = procstat_getvmmap(psp, kipp, &vmcnt);
+	if (kivp == NULL)
+		err(EX_OSERR, "procstat_getvmmap");
+
+	found_self = found_parent = false;
+	for (u_int i = 0; i < vmcnt; i++) {
+		if (kivp[i].kve_pid == pid)
+			found_self = true;
+		if (kivp[i].kve_pid == ppid)
+			found_parent = true;
+	}
+	if (!found_self)
+		errx(EX_SOFTWARE, "Didn't find self in vmstate");
+
+	procstat_freevmmap(psp, kivp);
+	procstat_freeprocs(psp, kipp);
+	procstat_close(psp);
+	return (found_parent);
+}
+
+#ifdef CHERIBSD_DYNAMIC_TESTS
```

Comment:
Why is this test dynamic only?  Shouldn't it also work for static binaries?

Reply:
This check forbids static binaries.: https://github.com/CTSRD-CHERI/cheribsd/blob/069ade2f9c5eb1ea945fd44f37a2b9246dc62785/sys/kern/imgact_elf.c#L1502-L1511
I'm not convinced this is the right balance and suspect that PIE static binaries should be allowed.

Reply:
Oh, I think FreeBSD doesn't yet support PIE static binaries (you need a small bit of ld.so in crt to deal with relocating the binary to support static PIE and we don't have that).

Reply:
I've updated this bit of code to be more clear what's going on and restore the uprintf in the default case.